cdc/generation.cc

```diff
@@ -59,14 +59,57 @@ static void copy_int_to_bytes(int64_t i, size_t offset, bytes& b) {
     std::copy_n(reinterpret_cast<int8_t*>(&i), sizeof(int64_t), b.begin() + offset);
 }
 
-stream_id::stream_id(int64_t first, int64_t second)
+static constexpr auto stream_id_version_bits = 4;
+static constexpr auto stream_id_random_bits = 38;
+static constexpr auto stream_id_index_bits = sizeof(uint64_t)*8 - stream_id_version_bits - stream_id_random_bits;
+
+static constexpr auto stream_id_version_shift = 0;
+static constexpr auto stream_id_index_shift = stream_id_version_shift + stream_id_version_bits;
+static constexpr auto stream_id_random_shift = stream_id_index_shift + stream_id_index_bits;
+
+/**
+ * Responsibilty for encoding stream_id moved from factory method to 
+ * this constructor, to keep knowledge of composition in a single place. 
+ * Note this is private and friended to topology_description_generator, 
+ * because he is the one who defined the "order" we view vnodes etc. 
+ */ 
+stream_id::stream_id(dht::token token, size_t vnode_index)
     : _value(bytes::initialized_later(), 2 * sizeof(int64_t))
 {
-    copy_int_to_bytes(first, 0, _value);
-    copy_int_to_bytes(second, sizeof(int64_t), _value);
+    static thread_local std::mt19937_64 rand_gen(std::random_device{}());
+    static thread_local std::uniform_int_distribution<uint64_t> rand_dist;
+
+    auto rand = rand_dist(rand_gen);
+    auto mask_shift = [](uint64_t val, size_t bits, size_t shift) {
+        return (val & ((1ull << bits) - 1u)) << shift;
+    };
+    /**
+     *  Low qword:
+     * 0-4: version
+     * 5-26: vnode index as when created (see generation below). This excludes shards
+     * 27-64: random value (maybe to be replaced with timestamp)
+     */
+    auto low_qword = mask_shift(version_1, stream_id_version_bits, stream_id_version_shift)
+        | mask_shift(vnode_index, stream_id_index_bits, stream_id_index_shift)
+        | mask_shift(rand, stream_id_random_bits, stream_id_random_shift)
+        ;
+
+    copy_int_to_bytes(dht::token::to_int64(token), 0, _value);
```

Comment:
I think this is in a wrong order. low_qword should start from 0 and token from sizeof(int64_t).

Reply:
No, token is in high qword, which starts at byte offset 0, because we compose the byte array <high>:<low>. This ensures id:s are sorted in token order. Or to put it simply, token has not moved from original version. But low order qword is now <rand>:<index>:<ver>, with ver holding the lowest four bits in the whole 128 bits.

Reply:
Does this order matter at all?

Reply:
Yes and no. By having token (in network byte order) in the high qword of the byte array, we will effectively be sorting as if this was a tuple/uuid <token>:<low stuff>, which in turn means that stream ID:s are in token order.
This sort of matters at least in alternator, where we want to identify token ranges between generations to emulate the "resharding" in dynamo etc. Could of course just re-sort based on token, but imho all code is currently written using an implicit assumption that token order == id order. 
Also, remember, it was you who said "version" should be in lowest (first) 4 bits. That is exactly where they are now.

Reply:
I believed I said first 4 bits and I meant by this bits no. 0, 1, 2, and 3 in a blob. It felt more natural to me but if alternator needs the order to be preserved then I guess there's no way around it. Unless we do <version><token><rest> but not sure it makes sense

Reply:
It does not "need" it, since itcan always be sorted with a custom order. My point was just that it was at least one of the places where having token as msb imho makes sense. And I think it does in general as well. 
In fact, even from a debugging standpoint it makes sense, since it is much easier to see version that is `qwords{1]&0xf` compared to `(qword[0] >> 60)&0xf` when looking at a memdump or whatever. For me at least.

Reply:
> I meant by this bits no. 0, 1, 2, and 3 in a blob

Yeah, but which are those? Does `0` mean the most significant or the least significant bit?

Right now the version is the first 4 least significant bits, and it makes sense to me. When we make the ID longer we would extend it "to the left" (add more significant bits).

Reply:
I know it is equivalent to have the first 4 bits or the last 4 bits being the version. It just feels weird. How many protocols do you know that first put a message and then the type or version of the message?

Regarding the order, I wouldn't like to depend on that. In the future we may have totally different representation which may sort totally differently

I won't insist though as long as the version is on any of the ends to be easily accessible.